lightning/src/ln/channelmanager.rs

```diff
@@ -7906,6 +7704,346 @@ where
+		// could discover the final destination of X, by probing the adjacent nodes on the route
+		// with a keysend payment of identical payment hash to X and observing the processing
+		// time discrepancies due to a hash collision with X.
+		let hashed_preimage = PaymentHash(Sha256::hash(&payment_preimage.0).into_inner());
+		if hashed_preimage != payment_hash {
+			return Err(InboundOnionErr {
+				err_code: 0x4000|22,
+				err_data: Vec::new(),
+				msg: "Payment preimage didn't match payment hash",
+			});
+		}
+		if !accept_mpp_keysend && payment_data.is_some() {
+			return Err(InboundOnionErr {
+				err_code: 0x4000|22,
+				err_data: Vec::new(),
+				msg: "We don't support MPP keysend payments",
+			});
+		}
+		PendingHTLCRouting::ReceiveKeysend {
+			payment_data,
+			payment_preimage,
+			payment_metadata,
+			incoming_cltv_expiry: outgoing_cltv_value,
+			custom_tlvs,
+		}
+	} else if let Some(data) = payment_data {
+		PendingHTLCRouting::Receive {
+			payment_data: data,
+			payment_metadata,
+			incoming_cltv_expiry: outgoing_cltv_value,
+			phantom_shared_secret,
+			custom_tlvs,
+		}
+	} else {
+		return Err(InboundOnionErr {
+			err_code: 0x4000|0x2000|3,
+			err_data: Vec::new(),
+			msg: "We require payment_secrets",
+		});
+	};
+	Ok(PendingHTLCInfo {
+		routing,
+		payment_hash,
+		incoming_shared_secret: shared_secret,
+		incoming_amt_msat: Some(amt_msat),
+		outgoing_amt_msat: onion_amt_msat,
+		outgoing_cltv_value,
+		skimmed_fee_msat: counterparty_skimmed_fee_msat,
+	})
+}
+
+/// Peel one layer off an incoming onion, returning PendingHTLCInfo (either Forward or Receive).
+/// This does all the relevant context-free checks that LDK requires for payment relay or
+/// acceptance. If the payment is to be received, and the amount matches the expected amount for
+/// a given invoice, this indicates the UpdateAddHTLC, once fully committed in the channel, will
+/// generate an Event::PaymentReceived.
+pub fn peel_payment_onion<
+	NS: Deref,
+	L: Deref,
+	T: secp256k1::Verification,
+>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+	cur_height: u32, accept_mpp_keysend: bool,
+) -> Result<PendingHTLCInfo, InboundOnionErr>
+where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+{
+	let (hop, shared_secret, next_packet_details_opt) = 
+		decode_incoming_update_add_htlc_onion(msg, node_signer, logger, secp_ctx
+	).map_err(|_| InboundOnionErr {
+		msg: "Failed to decode update add htlc onion",
+		err_code: 0x4000 | 22,
+		err_data: Vec::new(),
+	})?;
+	Ok(match hop {
+		onion_utils::Hop::Forward{ next_hop_data, next_hop_hmac, new_packet_bytes } => {
+			if next_packet_details_opt.is_none() {
+				// Forward should always include the next hop details
+				return Err(InboundOnionErr {
+					msg: "Failed to decode update add htlc onion",
+					err_code: 0x4000 | 22,
+					err_data: Vec::new(),
+				});
+			};
+			let (next_packet_pubkey, _, _, outgoing_cltv_value) = next_packet_details_opt.unwrap();
```

Comment:
I prefer doing it this way to avoid an `unwrap`, even though it's safe (similar elsewhere):
```suggestion
			let (next_packet_pubkey, outgoing_cltv_value) = match next_packet_details_opt {
				Some((pk, _, _, cltv)) => (pk, cltv),
				None => {
					// Forward should always include the next hop details
					return Err(InboundOnionErr {
						msg: "Failed to decode update add htlc onion",
						err_code: 0x4000 | 22,
						err_data: Vec::new(),
					});
				}
			};
```